File: packages/cli/src/commands/doctor.ts
import { access } from 'node:fs/promises';
import { getConfig } from '../utils/config.js';
import {
  checkNetworkReachable,
  hasDockerPermissions,
  isDockerInstalled,
  isDockerRunning,
} from '../utils/docker.js';
import { verifyVolume } from '../utils/volume.js';
import { log } from '../utils/ui.js';

async function checkPath(path: string): Promise<boolean> {
  try {
    await access(path);
    return true;
  } catch {
    return false;
  }
}

function printCheck(name: string, ok: boolean, hint?: string): void {
  if (ok) {
    log.success(name);
  } else {
    log.error(name);
    if (hint) {
      log.info(`  ${hint}`);
    }
  }
}

export async function doctorCommand(): Promise<void> {
  const config = getConfig();
  const dockerInstalled = await isDockerInstalled();
  const dockerRunning = dockerInstalled ? await isDockerRunning() : false;
  const dockerPerms = dockerRunning ? await hasDockerPermissions() : false;

  printCheck('Docker installed', dockerInstalled, 'Install Docker from docs.docker.com/get-docker');
  printCheck('Docker daemon running', dockerRunning, 'Start Docker service');
  printCheck('Docker permissions', dockerPerms, 'Add user to docker group');

  const gpu = await checkPath('/dev/dri');
  const pulse = Boolean(process.env.XDG_RUNTIME_DIR) && (await checkPath(`${process.env.XDG_RUNTIME_DIR}/pulse/native`));
  const display = Boolean(process.env.DISPLAY || process.env.WAYLAND_DISPLAY);

  printCheck('GPU device available', gpu, 'GPU passthrough disabled if missing');
  printCheck('Audio socket available', pulse, 'PulseAudio may not be available');
  printCheck('Display environment available', display, 'Set DISPLAY or WAYLAND_DISPLAY');

  const volume = await verifyVolume(config.volumeName);
  printCheck('Volume integrity', volume.ok, volume.reason);

  const ghcr = await checkNetworkReachable('ghcr.io');
  printCheck('Network access to ghcr.io', ghcr, 'Check DNS/network connectivity');
}

File: packages/cli_native/src/commands/doctor.rs
use anyhow::Result;
use std::path::Path;

use crate::utils::{
    config::get_config,
    docker::{
        check_network_reachable, has_docker_permissions, is_docker_installed, is_docker_running,
    },
    ui::LOG,
    volume::verify_volume,
};

fn print_check(name: &str, ok: bool, hint: Option<&str>) {
    if ok {
        LOG.success(name);
    } else {
        LOG.error(name);
        if let Some(h) = hint {
            LOG.info(&format!("  {}", h));
        }
    }
}

pub async fn doctor_command() -> Result<()> {
    let config = get_config();

    let docker_installed = is_docker_installed().await;
    let docker_running = if docker_installed {
        is_docker_running().await
    } else {
        false
    };
    let docker_perms = if docker_running {
        has_docker_permissions().await
    } else {
        false
    };

    print_check(
        "Docker installed",
        docker_installed,
        Some("Install Docker from docs.docker.com/get-docker"),
    );
    print_check(
        "Docker daemon running",
        docker_running,
        Some("Start Docker service"),
    );
    print_check(
        "Docker permissions",
        docker_perms,
        Some("Add user to docker group: sudo usermod -aG docker $USER"),
    );

    let gpu = Path::new("/dev/dri").exists();
    let pulse_socket = std::env::var("XDG_RUNTIME_DIR")
        .map(|d| Path::new(&d).join("pulse/native").exists())
        .unwrap_or(false);
    let display_set = std::env::var("DISPLAY").is_ok() || std::env::var("WAYLAND_DISPLAY").is_ok();

    print_check(
        "GPU device available",
        gpu,
        Some("GPU passthrough disabled if missing"),
    );
    print_check(
        "Audio socket available",
        pulse_socket,
        Some("PulseAudio may not be available"),
    );
    print_check(
        "Display environment available",
        display_set,
        Some("Set DISPLAY or WAYLAND_DISPLAY"),
    );

    let volume_result = verify_volume(&config.volume_name).await;
    print_check(
        "Volume integrity",
        volume_result.ok,
        volume_result.reason.as_deref(),
    );

    let ghcr = check_network_reachable("ghcr.io").await;
    print_check(
        "Network access to ghcr.io",
        ghcr,
        Some("Check DNS/network connectivity"),
    );

    Ok(())
}
